### PR TITLE
Allow VerifiableCredentials in the oauth2 flow

### PR DESCRIPTION
**Pull request by dmeijboom**
Allow VerifiableCredentials in the oauth2 flow. If present they will be validated.

### CODE REVIEW

File: docs/_static/auth/v1.yaml [hunk 1]
```diff
@@ -279,6 +279,67 @@ paths:
 
 components:
   schemas:
+    VerifiableCredential:
+      type: object
+      description: A credential according to the W3C and Nuts specs.
+      required:
+        - "@context"
```
Comment (stevenvegt) [inline docs/_static/auth/v1.yaml:286]: Many of these fields can also contain a string when there is only 1 value.

**Review by stevenvegt — COMMENTED** [review]
If a token does not have credentials, it is valid? Feels like we might need a higher level DIDMan method which actually checks if the correct credentials are present. 🤔

File: docs/_static/auth/v1.yaml (hunk 1 above)
Comment (dmeijboom) [reply docs/_static/auth/v1.yaml:286]: Note that the `VerifiableCredential` type is a 1-on-1 copy from the `vcr` OpenAPI spec. Currently, this type is unused in Go as it's aliased to `vc.VerifiableCredential`. For now this feels like an "OK" solution.

**Comment by dmeijboom** [comment]
> If a token does not have credentials, it is valid? Feels like we might need a higher level DIDMan method which actually checks if the correct credentials are present. 🤔

It's valid without credentials indeed (they are optional in the spec).

File: auth/services/oauth/oauth.go [hunk 2]
```diff
@@ -273,6 +274,12 @@ func (s *service) CreateJwtGrant(request services.CreateJwtGrantRequest) (*servi
 		return nil, err
 	}
 
+	for _, verifiableCredential := range request.Credentials {
```
Comment (reinkrul) [inline auth/services/oauth/oauth.go:277]: I think these validation steps are missing (chapter 5.2.1.7 of RFC003):

> The credential issuer equals the sub field of the JWT.
> The credential credentialSubject.id equals the iss field of the JWT.

File: auth/services/oauth/oauth.go [hunk 3]
```diff
@@ -273,6 +274,12 @@ func (s *service) CreateJwtGrant(request services.CreateJwtGrantRequest) (*servi
...
+		if err := credential.Validate(verifiableCredential); err != nil {
+			return nil, fmt.Errorf("invalid VerifiableCredential: %w", err)
+		}
+	}
+
 	endpointID, _, err := services.ResolveCompoundServiceURL(s.docResolver, *custodian, request.Service, services.OAuthEndpointType, nil)
```
Comment (reinkrul) [inline auth/services/oauth/oauth.go:283]: `NutsAuthorizationCredential` specific part is not implemented yet?

> The name and city MUST exactly (case sensitive) match with a valid Verifiable Credential issued to the actor. This must be a NutsOrganizationCredential.

(RFC003 5.2.1.5 )

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: auth/services/oauth/oauth.go (hunk 2 above)
Comment (dmeijboom) [reply auth/services/oauth/oauth.go:277]: Note that this PR is based on: https://github.com/nuts-foundation/nuts-specification/pull/119 where any verifiable credential can be used. However, we could validate known types (such as `NutsAuthorizationCredential`)?

File: auth/services/oauth/oauth.go (hunk 3 above)
Comment (dmeijboom) [reply auth/services/oauth/oauth.go:283]: See: https://github.com/nuts-foundation/nuts-specification/pull/119

Comment (reinkrul) [reply auth/services/oauth/oauth.go:283]: Ah right I missed that.

File: auth/services/oauth/oauth.go (hunk 2 above)
Comment (reinkrul) [reply auth/services/oauth/oauth.go:277]: Well it must be validated _somewhere_, maybe we can have specialised validators per credential type?

Comment (dmeijboom) [reply auth/services/oauth/oauth.go:277]: I think we already have something like that. I'll take a look!

**Review by reinkrul — APPROVED** [review]

**Pull request merged**